### PR TITLE
ICU-22984 Generate old monkeys

### PR DESCRIPTION
**Pull request by eggrobin**
C++ only for now, I will do the [old monkeys from Java](https://en.wikipedia.org/wiki/Javan_surili) separately (that will be more of the same but with subtle differences, as we will need to expand the UnicodeSets ourselves before feeding them to Pattern). This uses ICU regexes to match the context before and after, or to match the left-hand side of a remap rule, as in UAX14 and UAX29. The regexes are the ones used to generate the conformance tests in the Unicode tools; they now closely match the ones in the UAX.
Also line breaking only; the other ones will be very similar (and simpler).

The generated partition and rules are from https://github.com/unicode-org/unicodetools/pull/979. Note that the generation involves transforming `&` and `-` (UnicodeSet syntax) to `&&` and `--` (ICU regex character class syntax).

#### Checklist
- [x] Required: Issue filed: ICU-22984
- [x] Required: The PR title must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. Example: "ICU-1234 Fix xyz"
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Review by macchiati — APPROVED** [review]
Looks like a good optimization (and cleanup)

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [inline icu4c/source/test/intltest/rbbitst.cpp:2858]: I think that's because . matches CR LF, so it won't match just half of that.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [inline icu4c/source/test/intltest/rbbitst.cpp:3038]: pse check & fix indentations...

**Review by markusicu — COMMENTED** [review]
very nice replacement of hardcoded rules with generated ones!

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: I will do another pass to try to make more things signed. As you can see from the commit history, I made some of them unsigned to deal with (perhaps overzealous) signed-unsigned comparison warnings, and then realized that I had written code that was warning-free but also UB, so went back to int64_t…

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: Ah, I did not know about IcuTestErrorCode. I will look into that.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [reply icu4c/source/test/intltest/rbbitst.cpp]: Why not int32_t? None of the test strings will be >1B characters.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp:2858]: That may well be the explanation, but even with regex greed, that still feels like a bug: `/(rn|[a-z])*n/` still matches `rn`.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: I ended up just `std::terminate`ing with a message as I have done elsewhere, since bubbling up the error all the way to the test would make debugging tricky.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: Done. `>` should really never happen unless you throw some extremely evil rules at this though. Added a long comment to that effect.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [inline icu4c/source/test/intltest/rbbitst.cpp]: should the final offset be -2?

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [inline icu4c/source/test/intltest/rbbitst.cpp:1731]: Please change  if (!U_SUCCESS(status))  to if (U_FAILURE(status))  -- at least two of these

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (markusicu) [inline icu4c/source/test/intltest/rbbitst.cpp:1605]: In ICU, we normally use int32_t not int. In the old days, an int could be just 16 bits wide. I suppose it could become 64 bits on some machine. Anyway, the functions involved explicitly return int32_t, so that would guarantee a type match.

**Review by markusicu — COMMENTED** [review]
great comments!

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: No, the `offset` variable is with respect to the previous value of indexInRemapped, so 6 + -1 + -1 = 4.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp:1731]: https://www.youtube.com/watch?v=vC7C4BEvhRs

Done.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp:1605]: This file is full of `int`s, but done.

(But then if we are not going with the concision of `int`, why not `int64_t`? It’s not like there is a performance benefit these days, and we did [get bitten something being 32-bit here](https://github.com/unicode-org/icu/commit/cc64ec7c94a822ae97dea5cd5a4fa972b2e465ef)…)

Comment (markusicu) [reply icu4c/source/test/intltest/rbbitst.cpp]: Where 32 bits are plenty, I still prefer to use them. It's also consistent with long-standing ICU practice, and consistent with Java whose "int" is precisely int32_t, and is used for string indexes etc., so can help a little with porting between the languages.

Comment (markusicu) [reply icu4c/source/test/intltest/rbbitst.cpp]: ... and we universally use `int32_t` not `std::int32_t` but I won't block the PR on that now...

**Review by markusicu — APPROVED** [review]

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (eggrobin) [reply icu4c/source/test/intltest/rbbitst.cpp]: > ... and we universally use int32_t not std::int32_t but I won't block the PR on that now...

But I did it anyway, so now you get to restamp the rubber :-p

**Review by markusicu — APPROVED** [review]

**Comment by markusicu** [comment]
Remember to squash the commits. It's ok if you want to merge multiple commits if they make sense as separate ones.
Each one's message needs to start with `ICU-22984 `
Squashing without contents change will not require re-approval.

**Comment by eggrobin** [comment]
> It's ok if you want to merge multiple commits if they make sense as separate ones.

I tried with four steps (unicodetools-like implementation, optimization, code motion which wants to be on its own because it is undiffable, post-motion cleanup). Let me know if you like that, otherwise I can squish it all into one commit.

**Comment by markusicu** [comment]
> > It's ok if you want to merge multiple commits if they make sense as separate ones.
> 
> I tried with four steps (unicodetools-like implementation, optimization, code motion which wants to be on its own because it is undiffable, post-motion cleanup). Let me know if you like that, otherwise I can squish it all into one commit.

wfm tnx
I added the allow-many-commits magic into the description.

**Comment by FrankYFTang** [comment]
The Exhaustive Tests for ICU #22  is broken between e025466 and  2e57f07 in TestMonkey

https://github.com/unicode-org/icu/actions/runs/12209763924 

Likely caused by this PR

**Comment by eggrobin** [comment]
Yes, fixing in #3296.

**Pull request merged**
